### ISSUE TITLE
Correctness: BlockGraph mis-models nested try/finally leave (inner endfinally bypasses enclosing finally)

### ISSUE THREAD

**Issue by richlander** [richlander/dotnet-inspect#2016]
## Summary

`ILInspector.Instructions.BlockGraph` mis-models the control flow of a `leave`
that exits **nested** `try/finally` regions: the inner `endfinally` is wired
directly to the *ultimate* `leave` target, **bypassing the enclosing finally
handler(s)**. Per ECMA-335, a `leave` to a target outside multiple protected
regions must execute each enclosing `finally` innermost-to-outermost before
transferring control. The current edges skip the outer finally entirely.

This is a **latent substrate CFG-fidelity bug** (in #1982): it is *equivalent to
the old hand-rolled builders* (so it is NOT a regression and did not surface in
the RD cutover #1990 or the LibraryBodyIndex convergence #2013), but it produces
a **false control-flow path** for any consumer that reasons about "what happens
on every path." The first consumer to hit it is the Leak Triage spike (#2014),
where it causes a high-severity **false positive**.

Found by the #2014 confirming adversarial review (Gemini 3.1 Pro) and confirmed
empirically (below).

## The defect

`BlockGraph.AddExceptionEdges` (around `src/ILInspector.Instructions/BlockGraph.cs:293-310`):

```csharp
if (region.Kind == HandlerKind.Finally)
{
    var leaveTargets = LeaveTargetsLeaving(region, instructions).ToArray();
    ...
    foreach (int block in BlocksInRange(region.HandlerStart, region.HandlerEnd))
    {
        if (lastByBlock[block] is not { OpCode: ILOpCode.Endfinally }) continue;
        foreach (int target in leaveTargets)
            successorsByBlock[block].Add(targetBlock);   // inner endfinally -> ULTIMATE target
    }
}
```

For a `leave` that also exits an enclosing finally, the inner `endfinally`
should flow to the **next enclosing finally's entry**, not to the final target.
Only the outermost crossed finally's `endfinally` should flow to the `leave`
target.

## Empirical repro

```csharp
using System.Buffers;
public static class Repro
{
    public static void NestedLeaveFalsePositive()
    {
        var buffer = ArrayPool<byte>.Shared.Rent(16);
        try
        {
            try { goto L_Outer; }            // leave crosses both finallys
            finally { System.Console.WriteLine("Inner"); }
        }
        finally { ArrayPool<byte>.Shared.Return(buffer); }  // RETURN lives here
        L_Outer: return;
    }
}
```

Running the #2014 Leak Triage spike on the compiled assembly:

```
| Member | Shape | Severity |
| `Repro.NestedLeaveFalsePositive()` | arraypool-rent-not-returned | high |
```

The buffer IS returned (outer finally), so this is a false positive. The CFG
path `inner-try -> inner-finally -> L_Outer` skips the outer finally, so the
analyzer sees a path with no `Return`.

## Impact / scope

- **Correctness of `BlockGraph` exception edges** for nested `try/finally` with a
  region-crossing `leave`. Any all-paths/dataflow reasoning over such methods is
  affected.
- `ReachingDefinitions` (now on the product path via #1990) consumes these edges.
  The corpus no-op held for #1990/#2013 because this exact shape (nested finally +
  cross-region leave) is rare in the corpus libraries — but RD results could be
  subtly wrong for methods that do have it. **Any fix here must re-verify the
  #1974 corpus stays a no-op**, since the edges are now product-bound.
- Leak Triage (#2014) is the first consumer to surface it as a user-visible
  false positive.

## Proposed fix

Chain nested `endfinally` blocks: an `endfinally` whose `leave` target lies
outside an enclosing finally flows to that enclosing finally's entry (innermost
to outermost); only the outermost crossed finally flows to the actual target.
Mirror ECMA-335 `leave` semantics.

- Add a fixture covering single and multiply-nested finally + cross-region leave.
- Re-run the substrate fidelity gates and the #1974 corpus no-op (edges are now
  product-bound via RD).

## Interim

#2014 should fail-closed on nested-finally-crossing lifecycles (decline → no
finding) until this lands, so the spike stays precision-clean.

## Consult

This is substrate (#1982 / #1939) territory — requesting input from the
substrate author on whether to fix the edge chaining here vs adjust
`LeaveTargetsLeaving`, and on the corpus-reverification plan, before implementing.

Related: #1982, #1939, #1990, #2013, #2014, #1974.
